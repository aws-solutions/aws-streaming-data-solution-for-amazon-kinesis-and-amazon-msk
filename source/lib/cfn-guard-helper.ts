/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';

export class CfnGuardHelper {
    public static addSuppressions(resource: cdk.CfnResource, suppressions: string | string[]) {
        let rules = [];

        if (suppressions instanceof Array) {
            for (const suppression of suppressions) {
                rules.push(suppression);
            }
        } else {
            rules.push(suppressions);
        }

        // Ensure that the metadata property exists.
        if (!resource.cfnOptions.metadata) {
            resource.cfnOptions.metadata = {};
        }

        if (resource.cfnOptions.metadata.guard?.SuppressedRules) {
            // If the CfnResource already contains some suppressions, we don't want to erase them.
            const existingRules = resource.cfnOptions.metadata.guard.SuppressedRules;
            rules = [...existingRules, ...rules];
        }

        // It's possible that multiple constructs try to add the same suppression.
        // We only keep one occurrence (last) of each.
        const uniqueRules = [
            ...new Map(
                rules.map(rule => [rule, rule])
            ).values()
        ];

        resource.cfnOptions.metadata.guard = {
            SuppressedRules: uniqueRules
        }
    }
}
