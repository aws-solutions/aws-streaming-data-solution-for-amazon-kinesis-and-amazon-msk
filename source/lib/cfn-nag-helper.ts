/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';

export interface CfnNagSuppression {
    Id: string;
    Reason: string;
}

export class CfnNagHelper {
    public static addSuppressions(resource: cdk.CfnResource, suppressions: CfnNagSuppression | CfnNagSuppression[]) {
        let rules = [];

        if (suppressions instanceof Array) {
            for (const suppression of suppressions) {
                rules.push({ id: suppression.Id, reason: suppression.Reason });
            }
        } else {
            rules.push({ id: suppressions.Id, reason: suppressions.Reason });
        }

        if (resource.cfnOptions.metadata?.cfn_nag) {
            // If the CfnResource already contains some suppressions, we don't want to erase them.
            const existingRules = resource.cfnOptions.metadata.cfn_nag.rules_to_suppress;
            rules = [...existingRules, ...rules];
        }

        // It's possible that multiple constructs try to add the same suppression.
        // We only keep one occurrence (last) of each.
        // Based on https://stackoverflow.com/a/56768137
        const uniqueRules = [
            ...new Map(
                rules.map(rule => [rule.id, rule])
            ).values()
        ];

        if (!resource.cfnOptions.metadata) {
            resource.cfnOptions.metadata = {};
        }

        resource.cfnOptions.metadata.cfn_nag = {
            rules_to_suppress: uniqueRules
        }
    }
}
