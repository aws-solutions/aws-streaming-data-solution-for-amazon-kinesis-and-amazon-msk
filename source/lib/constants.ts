/*********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

export default class Constants {
    private static readonly ssmManagedActions = [
        'ssm:DescribeAssociation',
        'ssm:GetDeployablePatchSnapshotForInstance',
        'ssm:GetDocument',
        'ssm:DescribeDocument',
        'ssm:GetManifest',
        'ssm:GetParameter',
        'ssm:GetParameters',
        'ssm:ListAssociations',
        'ssm:ListInstanceAssociations',
        'ssm:PutInventory',
        'ssm:PutComplianceItems',
        'ssm:PutConfigurePackageResult',
        'ssm:UpdateAssociationStatus',
        'ssm:UpdateInstanceAssociationStatus',
        'ssm:UpdateInstanceInformation',
        'ec2messages:AcknowledgeMessage',
        'ec2messages:DeleteMessage',
        'ec2messages:FailMessage',
        'ec2messages:GetEndpoint',
        'ec2messages:GetMessages',
        'ec2messages:SendReply',
        'ssmmessages:CreateControlChannel',
        'ssmmessages:CreateDataChannel',
        'ssmmessages:OpenControlChannel',
        'ssmmessages:OpenDataChannel'
    ]

    public static get SsmManagedActions(): string[] {
        return Constants.ssmManagedActions;
    }
}