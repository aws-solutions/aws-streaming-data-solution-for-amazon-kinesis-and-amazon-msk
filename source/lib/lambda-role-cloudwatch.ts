/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */


import * as cdk  from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_iam as iam } from 'aws-cdk-lib';
import { CfnGuardHelper } from './cfn-guard-helper';
import { CfnRole } from 'aws-cdk-lib/aws-iam';

export interface ExecutionRoleProps {
    readonly inlinePolicyName: string;
    readonly inlinePolicyDocument: iam.PolicyDocument;
}

export class ExecutionRole extends Construct {
    public readonly Role: iam.IRole;

    constructor(scope: Construct, id: string, props?: ExecutionRoleProps) {
        super(scope, id);

        const logsPolicy = new iam.PolicyStatement({
            resources: [`arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:log-group:/aws/lambda/*`],
            actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents']
        });

        const inlinePolicies = {
            CloudWatchLogsPolicy: new iam.PolicyDocument({
                statements: [logsPolicy]
            })
        };

        if (props !== undefined) {
            (inlinePolicies as any)[props.inlinePolicyName] = props.inlinePolicyDocument;
        }

        this.Role = new iam.Role(this, 'Role', {
            assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
            inlinePolicies
        });

        // Add suppressions
        CfnGuardHelper.addSuppressions(this.Role.node.defaultChild as CfnRole, 'IAM_NO_INLINE_POLICY_CHECK');
    }
}