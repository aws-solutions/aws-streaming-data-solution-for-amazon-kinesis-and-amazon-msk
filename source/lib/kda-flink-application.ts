/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_kinesisanalytics as analytics, aws_iam as iam } from 'aws-cdk-lib';

import { CfnNagHelper } from './cfn-nag-helper';
import { CfnGuardHelper } from './cfn-guard-helper';
import { FlinkBase, FlinkBaseProps } from './kda-base';

export interface FlinkApplicationProps extends FlinkBaseProps {
    readonly environmentProperties: analytics.CfnApplicationV2.PropertyGroupProperty;
    readonly metricsLevel: string;

    readonly codeBucketArn: string;
    readonly codeFileKey: string;

    readonly enableSnapshots: string;
    readonly enableAutoScaling: string;
}

export class FlinkApplication extends FlinkBase {
    constructor(scope: Construct, id: string, props: FlinkApplicationProps) {
        super(scope, id, props);
        this.addCfnNagSuppressions();
        this.addCfnGuardSuppressions();
    }

    protected createApplication(props: FlinkApplicationProps): analytics.CfnApplicationV2 {
        const autoScalingCondition = new cdk.CfnCondition(this, 'EnableAutoScaling', {
            expression: cdk.Fn.conditionEquals(props.enableAutoScaling, 'true')
        });

        const snapshotCondition = new cdk.CfnCondition(this, 'EnableSnapshots', {
            expression: cdk.Fn.conditionEquals(props.enableSnapshots, 'true')
        });

        return new analytics.CfnApplicationV2(this, 'Application', {
            runtimeEnvironment: 'FLINK-1_15',
            serviceExecutionRole: this.Role.roleArn,
            applicationConfiguration: {
                applicationCodeConfiguration: {
                    codeContent: {
                        s3ContentLocation: {
                            bucketArn: props.codeBucketArn,
                            fileKey: props.codeFileKey
                        }
                    },
                    codeContentType: 'ZIPFILE'
                },
                environmentProperties: {
                    propertyGroups: [props.environmentProperties]
                },
                flinkApplicationConfiguration: {
                    monitoringConfiguration: {
                        configurationType: 'CUSTOM',
                        logLevel: props.logLevel,
                        metricsLevel: props.metricsLevel
                    },
                    parallelismConfiguration: {
                        configurationType: 'CUSTOM',
                        autoScalingEnabled: cdk.Fn.conditionIf(autoScalingCondition.logicalId, true, false)
                    },
                    checkpointConfiguration: {
                        configurationType: 'DEFAULT'
                    }
                },
                applicationSnapshotConfiguration: {
                    snapshotsEnabled: cdk.Fn.conditionIf(snapshotCondition.logicalId, true, false)
                }
            }
        });
    }

    protected createRole(props: FlinkApplicationProps): iam.IRole {
        const s3CodePolicy = new iam.PolicyDocument({
            statements: [new iam.PolicyStatement({
                resources: [`${props.codeBucketArn}/${props.codeFileKey}`],
                actions: ['s3:GetObjectVersion', 's3:GetObject']
            })]
        });

        const logsPolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: [`arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:log-group:*`],
                    actions: ['logs:DescribeLogGroups']
                }),
                new iam.PolicyStatement({
                    resources: [this.LogGroup.logGroupArn],
                    actions: ['logs:DescribeLogStreams', 'logs:PutLogEvents']
                })
            ]
        });

        const vpcPolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: ['*'],
                    actions: [
                        'ec2:CreateNetworkInterface',
                        'ec2:DescribeNetworkInterfaces',
                        'ec2:DescribeVpcs',
                        'ec2:DeleteNetworkInterface',
                        'ec2:DescribeDhcpOptions',
                        'ec2:DescribeSubnets',
                        'ec2:DescribeSecurityGroups'
                    ]
                }),
                new iam.PolicyStatement({
                    resources: [`arn:${cdk.Aws.PARTITION}:ec2:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:network-interface/*`],
                    actions: ['ec2:CreateNetworkInterfacePermission']
                })
            ]
        });

        return new iam.Role(this, 'AppRole', {
            assumedBy: new iam.ServicePrincipal('kinesisanalytics.amazonaws.com'),
            inlinePolicies: {
                S3Policy: s3CodePolicy,
                LogsPolicy: logsPolicy,
                VpcPolicy: vpcPolicy
            }
        });
    }

    private addCfnNagSuppressions() {
        const cfnRole = this.Role.node.defaultChild as iam.CfnRole;
        CfnNagHelper.addSuppressions(cfnRole, {
            Id: 'W11',
            Reason: 'EC2 actions in VPC policy do not support resource level permissions'
        });
    }

    private addCfnGuardSuppressions() {
        const cfnRole = this.Role.node.defaultChild as iam.CfnRole;
        CfnGuardHelper.addSuppressions(cfnRole, ['IAM_NO_INLINE_POLICY_CHECK']);
    }
}
