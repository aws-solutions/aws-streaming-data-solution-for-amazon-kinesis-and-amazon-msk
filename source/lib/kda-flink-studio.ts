/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as analytics from '@aws-cdk/aws-kinesisanalytics';
import * as iam from '@aws-cdk/aws-iam';
import * as glue from '@aws-cdk/aws-glue';

import { CfnNagHelper } from './cfn-nag-helper';
import { FlinkBase, FlinkBaseProps } from './kda-base';

export class FlinkStudio extends FlinkBase {
    private DatabaseName: string = '';

    constructor(scope: cdk.Construct, id: string, props: FlinkBaseProps) {
        super(scope, id, props);
        this.addCfnNagSuppressions();
    }

    protected createRole(_props: FlinkBaseProps): iam.IRole {
        const glueDb = new glue.CfnDatabase(this, 'Database', {
            catalogId: cdk.Aws.ACCOUNT_ID,
            databaseInput: {
                description: `${cdk.Aws.STACK_NAME} - Database for Amazon Kinesis Data Analytics Studio`
            }
        });

        this.DatabaseName = glueDb.ref;

        const logsPolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: [`arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:log-group:*`],
                    actions: ['logs:DescribeLogGroups']
                }),
                new iam.PolicyStatement({
                    resources: [this.LogGroup.logGroupArn],
                    actions: ['logs:DescribeLogStreams', 'logs:PutLogEvents']
                })
            ]
        });

        const vpcPolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: ['*'],
                    actions: [
                        'ec2:CreateNetworkInterface',
                        'ec2:DescribeNetworkInterfaces',
                        'ec2:DescribeVpcs',
                        'ec2:DeleteNetworkInterface',
                        'ec2:DescribeDhcpOptions',
                        'ec2:DescribeSubnets',
                        'ec2:DescribeSecurityGroups'
                    ]
                }),
                new iam.PolicyStatement({
                    resources: [`arn:${cdk.Aws.PARTITION}:ec2:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:network-interface/*`],
                    actions: ['ec2:CreateNetworkInterfacePermission']
                })
            ]
        });

        // This policy is based on https://docs.aws.amazon.com/kinesisanalytics/latest/java/how-zeppelin-appendix-iam.html
        const gluePolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: [
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:connection/*`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:table/${this.DatabaseName}/*`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:database/${this.DatabaseName}`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:database/hive`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:catalog`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:userDefinedFunction/*`
                    ],
                    actions: [
                        'glue:GetConnection',
                        'glue:GetTable',
                        'glue:GetTables',
                        'glue:GetDatabase',
                        'glue:CreateTable',
                        'glue:UpdateTable',
                        'glue:GetUserDefinedFunction'
                    ]
                }),
                new iam.PolicyStatement({
                    resources: ['*'],
                    actions: ['glue:GetDatabases']
                }),
            ]
        });

        return new iam.Role(this, 'AppRole', {
            assumedBy: new iam.ServicePrincipal('kinesisanalytics.amazonaws.com'),
            inlinePolicies: {
                LogsPolicy: logsPolicy,
                VpcPolicy: vpcPolicy,
                GluePolicy: gluePolicy
            }
        });
    }

    protected createApplication(_props: FlinkBaseProps): analytics.CfnApplicationV2 {
        return new analytics.CfnApplicationV2(this, 'Studio', {
            runtimeEnvironment: 'ZEPPELIN-FLINK-2_0',
            applicationMode: 'INTERACTIVE',
            serviceExecutionRole: this.Role.roleArn,
            applicationConfiguration: {
                zeppelinApplicationConfiguration: {
                    catalogConfiguration: {
                        glueDataCatalogConfiguration: {
                            databaseArn: `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:database/${this.DatabaseName}`
                        }
                    },
                    monitoringConfiguration: {
                        logLevel: _props.logLevel
                    },
                    customArtifactsConfiguration: [
                        {
                            artifactType: 'DEPENDENCY_JAR',
                            mavenReference: {
                                groupId: 'org.apache.flink',
                                artifactId: 'flink-sql-connector-kinesis_2.12',
                                version: '1.13.2'
                            }
                        },
                        {
                            artifactType: 'DEPENDENCY_JAR',
                            mavenReference: {
                                groupId: 'org.apache.flink',
                                artifactId: 'flink-connector-kafka_2.12',
                                version: '1.13.2'
                            }
                        },
                        {
                            artifactType: 'DEPENDENCY_JAR',
                            mavenReference: {
                                groupId: 'software.amazon.msk',
                                artifactId: 'aws-msk-iam-auth',
                                version: '1.1.0'
                            }
                        }
                    ]
                }
            }
        });
    }

    private addCfnNagSuppressions() {
        const cfnRole = this.Role.node.defaultChild as iam.CfnRole;
        CfnNagHelper.addSuppressions(cfnRole, {
            Id: 'W11',
            Reason: 'EC2 actions do not support resource level permissions / Studio uses default Glue database'
        });
    }
}
