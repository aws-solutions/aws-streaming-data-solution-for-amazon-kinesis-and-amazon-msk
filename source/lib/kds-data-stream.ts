/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */


import * as cdk  from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_kinesis as kinesis, aws_lambda as lambda, aws_iam as iam } from 'aws-cdk-lib';

import { ExecutionRole } from './lambda-role-cloudwatch';
import { CfnNagHelper } from './cfn-nag-helper';

export interface DataStreamProps {
    readonly shardCount: number;
    readonly retentionPeriod: cdk.Duration;
    readonly enableEnhancedMonitoring: string;
}

export class DataStream extends Construct {
    public readonly Stream: kinesis.Stream;

    constructor(scope: Construct, id: string, props: DataStreamProps) {
        super(scope, id);

        if (!cdk.Token.isUnresolved(props.shardCount) && props.shardCount <= 0) {
            throw new Error('shardCount must be a positive number');
        }

        this.Stream = new kinesis.Stream(this, 'DataStream', {
            encryption: kinesis.StreamEncryption.MANAGED,
            shardCount: props.shardCount,
            retentionPeriod: props.retentionPeriod
        });

        this.createCustomResource(props.enableEnhancedMonitoring);
    }

    private createCustomResource(enableEnhancedMonitoring: string) {
        const customResouceRole = new ExecutionRole(this, 'Role', {
            inlinePolicyName: 'MonitoringPolicy',
            inlinePolicyDocument: new iam.PolicyDocument({
                statements: [new iam.PolicyStatement({
                    resources: ['*'],
                    actions: ['kinesis:EnableEnhancedMonitoring', 'kinesis:DisableEnhancedMonitoring']
                })]
            })
        });

        const cfnRole = customResouceRole.Role.node.defaultChild as iam.CfnRole;
        CfnNagHelper.addSuppressions(cfnRole, {
            Id: 'W11',
            Reason: 'Kinesis enhanced monitoring actions do not support resource level permissions'
        });

        const customResourceFunction = new lambda.Function(this, 'CustomResource', {
            runtime: lambda.Runtime.PYTHON_3_12,
            handler: 'lambda_function.handler',
            role: customResouceRole.Role,
            code: lambda.Code.fromAsset('lambda/kds-enhanced-monitoring'),
            timeout: cdk.Duration.seconds(30)
        });

        new cdk.CustomResource(this, 'EnhancedMonitoring', {
            serviceToken: customResourceFunction.functionArn,
            properties: {
                'EnableEnhancedMonitoring': enableEnhancedMonitoring,
                'StreamName': this.Stream.streamName
            },
            resourceType: 'Custom::EnhancedMonitoring'
        });
    }
}
