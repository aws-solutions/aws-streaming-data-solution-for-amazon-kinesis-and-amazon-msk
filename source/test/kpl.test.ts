/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */


import * as cdk  from 'aws-cdk-lib';
import { aws_kinesis as kinesis } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { KinesisProducer } from '../lib/kpl-producer';

let stack: cdk.Stack;
let testStream: kinesis.Stream;

beforeEach(() => {
    const app = new cdk.App();
    stack = new cdk.Stack(app, 'TestStack');
    testStream = new kinesis.Stream(stack, 'TestStream');
});

test('creates a KPL instance', () => {
    const kpl = new KinesisProducer(stack, 'TestProducer', {
        stream: testStream,
        vpcId: 'vpc-123',
        subnetId: 'subnet-abc',
        imageId: 'ami-1234',
        codeBucketName: 'test-bucket',
        codeFileKey: 'test-key.zip'
    });

    expect(kpl.InstanceId).not.toBeUndefined();
});

test('adds cfn_nag suppressions', () => {
    new KinesisProducer(stack, 'TestProducer', {
        stream: testStream,
        vpcId: 'vpc-123',
        subnetId: 'subnet-abc',
        imageId: 'ami-1234',
        codeBucketName: 'test-bucket',
        codeFileKey: 'test-key.zip'
    });

    Template.fromStack(stack).hasResource('AWS::IAM::Policy', {
        Metadata: {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W12',
                    reason: 'Session Manager actions do not support resource level permissions'
                }]
            }
        }
    });

    Template.fromStack(stack).hasResource('AWS::IAM::Policy', {
        Metadata: {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W12',
                    reason: 'PutMetricData action does not support resource level permissions'
                }]
            }
        }
    });

    Template.fromStack(stack).hasResource('AWS::EC2::SecurityGroup', {
        Metadata: {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W5',
                    reason: 'Outbound access is allowed to connect to Kinesis'
                }]
            }
        }
    });
});
