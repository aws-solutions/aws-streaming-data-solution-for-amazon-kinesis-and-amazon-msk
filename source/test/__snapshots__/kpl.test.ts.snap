// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a KPL instance 1`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Resources": Object {
    "TestProducer63E36A21": Object {
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "TestProducerInstanceProfile145E74DF",
        },
        "ImageId": "ami-1234",
        "InstanceType": "t3.small",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "TestProducerSecurityGroup159E2ABD",
              "GroupId",
            ],
          },
        ],
        "SubnetId": "subnet-abc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "KplInstance",
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash
yum update -y
amazon-linux-extras install java-openjdk11 -y
java -version
aws s3 cp s3://test-bucket/test-key.zip /tmp/producer.zip
unzip /tmp/producer.zip -d /tmp && rm -f /tmp/producer.zip",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "TestProducerCodePolicyC6F00C96": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObjectVersion",
                "s3:GetObject",
                "s3:ListBucket",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::test-bucket/test-key.zip",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::aws-bigdata-blog/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::aws-bigdata-blog",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestProducerCodePolicyC6F00C96",
        "Roles": Array [
          Object {
            "Ref": "TestProducerRoleD68492B6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestProducerInstanceProfile145E74DF": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "TestProducerRoleD68492B6",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "TestProducerMonitoringPolicy7F2E8F61": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "PutMetricData action does not support resource level permissions",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestProducerMonitoringPolicy7F2E8F61",
        "Roles": Array [
          Object {
            "Ref": "TestProducerRoleD68492B6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestProducerRoleD68492B6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestProducerRoleDefaultPolicy73CAADAF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:ListShards",
                "kinesis:PutRecord",
                "kinesis:PutRecords",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestStreamE6F40222",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "kinesis:DescribeStream",
                "kinesis:DescribeStreamSummary",
                "kinesis:DescribeStreamConsumer",
                "kinesis:RegisterStreamConsumer",
                "kinesis:SubscribeToShard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestStreamE6F40222",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestProducerRoleDefaultPolicy73CAADAF",
        "Roles": Array [
          Object {
            "Ref": "TestProducerRoleD68492B6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestProducerSecurityGroup159E2ABD": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W5",
              "reason": "Outbound access is allowed to connect to Kinesis",
            },
          ],
        },
      },
      "Properties": Object {
        "GroupDescription": "KPL security group",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow HTTP outbound traffic",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow HTTPS outbound traffic",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": "vpc-123",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestProducerSessionManagerPolicy1CCC3BB8": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Session Manager actions do not support resource level permissions",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ssm:UpdateInstanceInformation",
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestProducerSessionManagerPolicy1CCC3BB8",
        "Roles": Array [
          Object {
            "Ref": "TestProducerRoleD68492B6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestStreamE6F40222": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
        "StreamModeDetails": Object {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;
